peercrawler.py

```diff
@@ -217,15 +228,17 @@ def crawl_peer(peer: Peer):
             logger.debug("Query %41s:%5s (score:%4s)" % ('[%s]' % p.ip, p.port, p.score))
 
             new_peers = self.get_peers_from_peer(peer)
-            self.add_peers(new_peers)
+            self.add_peers(peer, new_peers)
 
         with ThreadPoolExecutor(max_workers=max_workers) as t:
             for p in peers_copy:
                 t.submit(crawl_peer, peer=p)
 
     def crawl(self, forever, delay, max_workers=4):
         initial_peers = get_all_dns_addresses_as_peers(self.ctx['peeraddr'], self.ctx['peerport'], -1)
-        self.add_peers(initial_peers)
+        with self.mutex:
```

Comment:
This is code duplication, why can't you use add_peers?

Reply:
add_peers takes in a parameter called from_peer, which is a single Peer object, and a collection of peers which contain peers that are connected to from_peer, while this code right here just adds a bunch of "independent" peers.
I'll try to improve it a bit though